restapi/src/main/java/io/stargate/web/docsapi/service/JsonDocumentShredder.java

```diff
@@ -0,0 +1,211 @@
+      if (idx >= config.getMaxArrayLength()) {
+        throw new ErrorCodeRuntimeException(ErrorCode.DOCS_API_GENERAL_ARRAY_LENGTH_EXCEEDED);
+      }
+
+      // convert the array index into path
+      // then create new next row builder
+      String arrayPath = "[" + DocsApiUtils.leftPadTo6(String.valueOf(idx)) + "]";
+      Supplier<ImmutableJsonShreddedRow.Builder> nextRowBuilder =
+          () -> rowBuilder.get().addPath(arrayPath);
+
+      // process inner node and increase the index
+      processNode(inner, nextRowBuilder, numericBooleans, result);
+      idx++;
+    }
+  }
+
+  private void processObjectNode(
+      JsonNode node,
+      Supplier<ImmutableJsonShreddedRow.Builder> rowBuilder,
+      boolean numericBooleans,
+      List<JsonShreddedRow> result) {
+    // empty object, simply create a reference to empty node and return
+    if (node.isEmpty()) {
+      ImmutableJsonShreddedRow row =
+          rowBuilder.get().stringValue(DocsApiConstants.EMPTY_OBJECT_MARKER).build();
+      result.add(row);
+      return;
+    }
+
+    node.fields()
+        .forEachRemaining(
+            field -> {
+              String fieldName = field.getKey();
+
+              // check for valid field name
+              if (DocsApiUtils.containsIllegalSequences(fieldName)) {
+                String msg =
+                    String.format(
+                        "Array paths contained in square brackets, periods, single quotes, and backslash are not allowed in field names, invalid field %s",
+                        fieldName);
+                throw new ErrorCodeRuntimeException(
+                    ErrorCode.DOCS_API_GENERAL_INVALID_FIELD_NAME, msg);
+              }
+
+              // escape the field path
+              // then create new next row builder
+              String fieldPath = DocsApiUtils.convertEscapedCharacters(fieldName);
+              Supplier<ImmutableJsonShreddedRow.Builder> nextRowBuilder =
+                  () -> rowBuilder.get().addPath(fieldPath);
+
+              // process inner node and increase the index
+              processNode(field.getValue(), nextRowBuilder, numericBooleans, result);
+            });
+  }
+
+  private void processValueNode(
+      JsonNode node,
+      Supplier<ImmutableJsonShreddedRow.Builder> rowBuilder,
+      boolean numericBooleans,
+      List<JsonShreddedRow> result) {
+    ImmutableJsonShreddedRow.Builder builder = rowBuilder.get();
+
+    // depending on the value type set values
+    if (node.isBoolean()) {
+      builder.booleanValue(convertToBackendBooleanValue(node.asBoolean(), numericBooleans));
+    } else if (node.isNumber()) {
+      builder.doubleValue(node.asDouble());
+    } else if (!node.isNull()) {
+      builder.stringValue(node.asText());
+    }
+
+    // build and add to the results
+    ImmutableJsonShreddedRow row = builder.build();
+    result.add(row);
+  }
+
+  private Object convertToBackendBooleanValue(boolean value, boolean numericBooleans) {
```

Comment:
Does this conversion need to be done here? It's seems like this layer might not need to deal with this. Would it be possible to use `boolean` here and the conversion to the numeric type be done in the storage layer right before it's stored?

Reply:
Yea seems that this makes sense.. I think I will apply this fix as well, shredder should not care about how are the values stored in the DB. Thanks :+1: